compiler/rustc_type_ir/src/relate.rs

```diff
@@ -0,0 +1,709 @@
+use std::iter;
+
+use rustc_ast_ir::Mutability;
+use rustc_type_ir::error::{ExpectedFound, TypeError};
+use rustc_type_ir::fold::TypeFoldable;
+use rustc_type_ir::inherent::*;
+use rustc_type_ir::{self as ty, InferCtxtLike, Interner};
+use tracing::{debug, instrument};
+
+pub type RelateResult<I, T> = Result<T, TypeError<I>>;
+
+pub trait TypeRelation<Infcx, I = <Infcx as InferCtxtLike>::Interner>: Sized
+where
+    Infcx: InferCtxtLike<Interner = I>,
+    I: Interner,
+{
+    fn tcx(&self) -> I;
+
+    /// Returns a static string we can use for printouts.
+    fn tag(&self) -> &'static str;
```

Comment:
not necessarily for this PR. Remove this and just use `std::any::type_name` for debug printing instead